browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

```diff
@@ -0,0 +1,125 @@
+
+#include "brave/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.h"
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include "brave/browser/ai_chat/common/constants.h"
+#include "brave/browser/ai_chat/common/pref_names.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "components/prefs/pref_service.h"
+
+using ai_chat::mojom::CharacterType;
+using ai_chat::mojom::ConversationTurn;
+using ai_chat::mojom::ConversationTurnVisibility;
+
+AIChatUIPageHandler::AIChatUIPageHandler(
+    TabStripModel* tab_strip_model,
+    Profile* profile,
+    mojo::PendingReceiver<ai_chat::mojom::PageHandler> receiver)
+    : profile_(profile), receiver_(this, std::move(receiver)) {
+  DCHECK(tab_strip_model);
+  tab_strip_model->AddObserver(this);
+
+  auto* web_contents = tab_strip_model->GetActiveWebContents();
+  if (!web_contents) {
+    return;
+  }
+
+  active_chat_tab_helper_ = AIChatTabHelper::FromWebContents(web_contents);
+
+  chat_tab_helper_observation_.Observe(active_chat_tab_helper_);
+}
+
+AIChatUIPageHandler::~AIChatUIPageHandler() = default;
+
+void AIChatUIPageHandler::SetClientPage(
+    mojo::PendingRemote<ai_chat::mojom::ChatUIPage> page) {
+  page_.Bind(std::move(page));
+}
+
+void AIChatUIPageHandler::SubmitHumanConversationEntry(
+    const std::string& input) {
+  // TODO(nullhook): Avoid copy
+  std::string input_copy = input;
+
+  // Prevent indirect prompt injections being sent to the AI model.
+  // TODO(nullhook): Abstract prompt injection cleanups to a central place
+  base::ReplaceSubstringsAfterOffset(&input_copy, 0, ai_chat::kHumanPrompt, "");
+  base::ReplaceSubstringsAfterOffset(&input_copy, 0, ai_chat::kAIPrompt, "");
+
+  active_chat_tab_helper_->MakeAPIRequestWithConversationHistoryUpdate(
+      {CharacterType::HUMAN, ConversationTurnVisibility::VISIBLE, input_copy});
+}
+
+void AIChatUIPageHandler::GetConversationHistory(
+    GetConversationHistoryCallback callback) {
+  std::vector<ConversationTurn> history =
+      active_chat_tab_helper_->GetConversationHistory();
+
+  std::vector<ai_chat::mojom::ConversationTurnPtr> list;
+
+  // Remove conversations that are meant to be hidden from the user
+  auto new_end_it = std::remove_if(
+      history.begin(), history.end(), [](const ConversationTurn& turn) {
+        return turn.visibility == ConversationTurnVisibility::HIDDEN;
+      });
+
+  std::transform(history.begin(), new_end_it, std::back_inserter(list),
+                 [](const ConversationTurn& turn) { return turn.Clone(); });
+
+  std::move(callback).Run(std::move(list));
+}
+
+void AIChatUIPageHandler::RequestSummary() {
+  active_chat_tab_helper_->RequestSummary();
+}
+
+void AIChatUIPageHandler::MarkAgreementAccepted() {
+  profile_->GetOriginalProfile()->GetPrefs()->SetBoolean(
+      ai_chat::prefs::kBraveChatHasSeenDisclaimer, true);
+}
+
+void AIChatUIPageHandler::OnPageChanged() {
+  page_.get()->OnConversationHistoryUpdate();
```

Comment:
This seems to be splitting the knowledge about what is happening with the data between the TabHelper and this PageHandler. In other words, shouldn't only 1 place know that when the page is changed, history is cleared (and that it happens before this `OnPageChanged` is called)? Isn't this already handled via the `OnHistoryUpdate` event forwarded from the TabHelper?
Also, why do we send OnGeneratingContent here when nothing is being generated? Is this to reset it always to false in case the page is changed in the middle of a chat API request for the previous page? If so, perhaps this can also be simply forwarded from the TabHelper, and the logic be contained there. An Android client would use the TabHelper, but not this PageHandler, so we need to keep it as dumb and specific to the WebUI as possible, and all the logic around the data and associated events contained in the TabHelper.

Reply:
`OnPageChanged` is triggered when the primary page is changed (to reset data) and when the active tab is changed (to switch data). Although I tried to keep it generic, it seems that this was not well communicated. 

I think `OnGeneratingContent` is not an appropriate name and has been changed. It's purpose is to reflect each tab helper's API progress to help display a loading indicator on the UI. 

If you notice in the updated code, `OnHistoryUpdate` is called when the active tab changes. This handles the data switching for conversation history and loading indicator. I think this should be part of the page handler, as this logic is related to the 'container' of the tab helpers. This keeps the tab helper straightforward with no state change surprises.

Reply:
Additionally, I have added a TODO to handle the cancellation of inflight async requests when the primary page changes. This should be part of the cleanup process, which involves resetting the data. However, I am not sure if this is a priority for this PR.

Reply:
At least part of my original point still stands - the Page Handler has this:

```
void AIChatUIPageHandler::OnHistoryUpdate() {
  page_.get()->OnConversationHistoryUpdate();
  page_.get()->OnAPIRequestInProgress(
      active_chat_tab_helper_->IsRequestInProgress());
}
```

which means this class needs to know that when the history is updated, then an API request happens to be in progress. And what I'm trying to say is that should be the job of a single place (the TabHelper) to know. We have baked that in to the WebUI class, and we'll have to bake that in to the Android / iOS UI. And the confusing thing is we already have a OnAPIRequestInProgress event, so why not just rely on TabHelper to send that through?

Reply:
Fixed. There shouldn't be any hidden calls now and mostly a 1:1 mapping between the PageHandler and the TabHelper